capsules/extra/src/nonvolatile_storage_driver.rs

```diff
@@ -320,6 +1044,34 @@ impl<'a> NonvolatileStorage<'a> {
                         .unwrap_or_else(|err| Err(err.into()))
                 })
             }
+            NonvolatileCommand::HeaderRead(_) | NonvolatileCommand::HeaderWrite(_) => {
+                self.header_buffer
+                    .take()
+                    .map_or(Err(ErrorCode::NOMEM), |header_buffer| {
+                        let active_len = cmp::min(length, header_buffer.len());
+
+                        // Check if there is something going on.
+                        if self.current_user.is_none() {
+                            match command {
+                                NonvolatileCommand::HeaderRead(action) => {
+                                    self.current_user.set(NonvolatileUser::HeaderManager(
+                                        HeaderState::Read(action),
+                                    ));
+                                    self.driver.read(header_buffer, offset, active_len)
+                                }
+                                NonvolatileCommand::HeaderWrite(action) => {
+                                    self.current_user.set(NonvolatileUser::HeaderManager(
+                                        HeaderState::Write(action),
+                                    ));
+                                    self.driver.write(header_buffer, offset, active_len)
```

Comment:
So this writes the header, which is constructed in ` fn write_region_header()`. I guess that ends up calling `NonvolatileToPages` `write()`, but where doe the erase happen?

The value was previously 0 and now we are trying to set it, but I can't figure out where the erase happens. What am I missing here?

Reply:
`NonvolatileStorage` allows us to just write.

Reply:
Why not just not re-allocate regions, at least for now?

Re-allocating implies deciding when an application is "done" with a region, and that is an entirely different question we don't have an answer for. I'm not saying we should never address that question, but it doesn't seem important in the short term.

Reply:
Yeah right now there is no concept of an app being "done" or freeing a region. The only way regions are re-used is if the magic header value is corrupted (which probably happens if the entire flash region is erased with some external tool from a computer (ex: nrfjprog)).

So maybe it's ok to just hand out regions without wiping them?

Reply:
> So maybe it's ok to just hand out regions without wiping them?

I think so. The magic value and id are the record of the storage ownership; if those are corrupted, well, what can you do? We have to store that information _somewhere_ and then need to expect it remains.

Reply:
> `NonvolatileStorage` allows us to just write.

But where does the erase happen? I don't see it in the code

Reply:
In whatever implements the interface.

The `NonvolatileStorage` API is generic across underlying memory technologies, i.e. FRAM or Flash can both implement `NonvolatileStorage`. If the layer beneath is a flash layer that requires something like erase-before-write, then the underlying layer is responsible for doing the erase before doing the write.

Reply:
The interface doesn't either, it's a bug in `NonvolatileToPages` and the nRF flash implementation. I'll open an issue